rfcs/0019-storing-client-data.md

```diff
@@ -0,0 +1,490 @@
+
+A drawback with sorting blocks and transaction based on the time that they are first seen is that if the client wants to store an arbitrary transaction and block, then they would not have the extra burden of tagging these objects with a timestamp for pagination. `block_compare` alleviates this issue.
+
+## SQL Database
+
+This architecture supports many SQL databases and it can even exclude a database entirely to make the architecture modular. When it is included, there are different use cases for using a light database, such as SQLite, and heavier database, such as Postgres. Postgres or even a distributed cloud database, such as AppSync, should be used when we have an archive node or a block explorer trying to make efficient writes to store everything that it hears from the network. These databases also support built-in subscriptions that would be convinient for clients to listen to.
+
+For the objects discussed in the previous section, we can embed them as tables in one global SQL database. Below are the schemas of the tables:
+
+```
+Table block {
+  state_hash string [pk]
+  creator string [not null]
+  staged_ledger_hash string [not null]
+  ledger_hash string [not null]
+  epoch int [not null]
+  slot int [not null]
+  ledger_proof_nonce int [not null]
+  status int [not null]
+  block_length int [not null]
+  block_time date [not null]
+  Indexes {
+    (block_length, epoch, slot) [name: "block_compare"]
+    block_time [name: "block_compare"]
+  }
+}
+
+  // Consensus_status is an ADT, but with its small state spaces, it can be coerced into an integer to make it easy to store it in a database. Here are the following states:
+  // - pending : The block is in the transition frontier and its value is its block confirmation number. Value[0, (k-1)]
+  // - confirmed : The block has reached finality by passing the root. Value=-1
+  // - failure: The block got removed from the transition frontier. Value=-2
+  // - unknown: After becoming online and bootstrapping, we may not know what will happen to the block. Value=-3
+
+Table user_commands {
+  id string [pk]
+  is_delegation bool [not null]
+  nonce int64 [not null]
+  sender string [not null]
+  receiver string [not null]
+  amount int64 [not null]
+  fee int64 [not null]
+  memo string [not null]
+  first_seen date
+  transaction_pool_membership bool [not null]
+  is_added_manually bool [not null]
+  Indexes {
+    (status, sender) [name: "fast_pooled_user_command_sender"]
+    (status, receiver) [name: "fast_pooled_user_command_receiver"]
+    (sender, first_seen) [name: "fast_sender_pagination"]
+    (receiver, first_seen) [name: "fast_receiver_pagination"]
+  }
+}
+
+Table fee_transfers {
```

Comment:
The table for fee_transfers and user_commands are different types. Fee_transfers do not have a sender column so it can be wasteful. The coinbase can be easily featured from the `block` Table as it has the field coinbase reward and we know the receiver of the reward, which is the proposer of the block.

Reply:
In general, I think it's more important to optimize for your relational model, not the fields you store. Having an extra NULL field on a table is not a great reason to make an extra relationship when the tables are otherwise related in the same fashion. I would still suggest consolidating the fee transfers and user commands. I buy the argument for coinbases though since there is only ever 1 coinbase per block and it's associated with the creator of that block.

Reply:
How does consolidating fee transfers and user commands into one object optimize database performance? Is it when you do a join? In the future, when we are going to add smart contracts and expand the functionality of the user commands, the user commands table will have more fields. This can be wasteful in space when we add fee transfers into this consolidated table. 

Overall I think having multiple tables would save space and not comprimise performance.